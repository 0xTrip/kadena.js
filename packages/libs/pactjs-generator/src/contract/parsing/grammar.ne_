@{%
  var { lexer, flattenRecursive, flattenOnce, wrap, filterFlattenWrap } = require('./lexer.js');
%}

@lexer lexer

main        -> _ ((%lparen _ %rparen | module | comment) _ ):+

# specific grammar
#  - module
module          -> %lparen _ %module _ name _ keysetOrGovernance _ meta _ moduleBody _ %rparen
keysetOrGovernance -> %atom
meta            -> ( doc _ | model _ ):*
doc             -> %string
                   | "@doc" _ %string
model           -> "@model" _ ( expression | listOfExpressions)
expression      -> %lparen _
                   ( maybeTypedAtom
                   | %decimal
                   | %symbol
                   | %boolean
                   | %string
                   | expression )
                   ( _ maybeTypedAtom
                   | _ %decimal
                   | _ %symbol
                   | _ %boolean
                   | _ %string
                   | _ expression):* _ %rparen
listOfExpressions -> %lsquare _ expression (_ expression):* _ %rsquare

moduleBody      -> ( _ ( comment
                     | defun
                     | defpact
                     | defcap
                     | deftable
                     | defschema
                     | defconst
                     | implements
                     | use
                     | bless
                   )):*
# - bless
implements      -> %lparen _ %implements _ name _ %rparen
# - defschema
defschema       -> %lparen _ %defschema _ name _ meta _ field:+ _ %rparen
field           -> _ maybeTypedAtom
# - deftable
deftable        -> %lparen _ %deftable _ maybeTypedSchema _ meta _ %rparen
maybeTypedSchema -> %atom (%colon %lcurly %atom %rcurly):?

#  - bless
bless           -> %lparen _ %bless _ %string _ %rparen
#  - defun
defun           -> %defun defunName argList defunBody
defunName       -> maybeTypedAtom
#  - defcap
defcap          -> %lparen _ %defcap _ name _ argList (_ doc):? _ body _ %rparen

# generic grammar
name            -> %atom
maybeTypedAtom  -> %atom
                   | typedAtom
typedAtom       -> %atom %colon %atom
argList         -> %lparen _ %rparen
                   | %lparen _ maybeTypedAtom ( _ %comma _ maybeTypedAtom ):* %rparen
body            -> ( expression (_ expression):* | bodyToken )
bodyToken -> %atom | %string | %boolean



#  | "[" _ value (_ "," _ value):* _ "]"

comment      -> %comment _

_            -> (%ws  | %nl  ):*
__           -> (%ws  | %nl  ):+
