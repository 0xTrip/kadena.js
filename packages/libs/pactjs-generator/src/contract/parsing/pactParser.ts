/* eslint-disable @kadena-dev/no-eslint-disable */
/* eslint-disable @rushstack/typedef-var */
/* eslint-disable @typescript-eslint/explicit-function-return-type */

import { unwrapData } from './utils/dataWrapper';
import { getCapabilities } from './utils/getCapabilities';
import { getBlockPointer, getPointer, IPointer } from './utils/getPointer';
import { functionCalls, parser } from './utils/pactGrammar';
import { FAILED } from './utils/parserUtilities';

export interface ISchema {
  name: string;
  doc?: string;
  properties?: Array<{
    name: string;
    type: string | { kind: string; value: string };
  }>;
}

interface IMethod {
  name: string;
  returnType?: string | { kind: string; value: string };
  doc?: string;
  parameters?: Array<{
    name: string;
    type: string | { kind: string; value: string };
  }>;
}

export interface IFunction extends IMethod {
  bodyPointer?: number;
  requiredCapabilities?: string[];
  withCapabilities?: string[];
  functionCalls?: {
    internal: string[];
    external: Array<{ namespace?: string; module: string; func: string }>;
  };
  externalFnCalls?: Array<{
    namespace?: string;
    module: string;
    func: string;
  }>;
  allExtractedCaps?: Array<{
    name: string;
    fullModuleName: string;
    reason: 'with-capability' | 'compose-capability';
    origin: string;
    capability: ICapability;
  }>;
}

export interface ICapability extends IMethod {
  composeCapabilities?: string[];
}

export interface IModule {
  kind: string;
  namespace: string;
  name: string;
  doc: string;
  governance: string;
  usedModules?: Array<{
    name: string;
    namespace?: string;
    hash?: string;
    imports?: string[];
  }>;
  usedInterface?: Array<{ name: string; namespace?: string }>;
  functions?: IFunction[];
  capabilities?: ICapability[];
  schemas?: ISchema[];
}

export interface IPactTree {
  namespace?: string[];
  usedModules?: Array<{ name: string; hash?: string }>;
  module?: IModule[];
}

// return the namespace of the module in case the file has multiple modules and namespaces
const getNamespace = (
  moduleLoc: number,
  allNamespaces?: Array<{ location: number; name: string }>,
) => {
  if (!allNamespaces) return '';
  const { name } = allNamespaces.reduce(
    (namespace, { location, name }) => {
      if (location < moduleLoc && location > namespace.location) {
        return { location, name };
      }
      return namespace;
    },
    { location: 0, name: '' },
  );
  return name;
};

// return the list of modules used in the module including the modules used in root level
const getUsedModules = (
  moduleLoc: number,
  allUsedModules?: Array<{
    location: number;
    name: string;
    namespace: string;
    hash?: string;
  }>,
) => {
  if (!allUsedModules) return [];
  const list = allUsedModules.reduce((list, { location, ...mod }) => {
    if (location < moduleLoc) {
      return [...list, mod];
    }
    return list;
  }, [] as Array<{ name: string; namespace: string; hash?: string; imports?: Array<string> }>);

  return list;
};

// returns the list of modules used in the functions of the module without using "use" keyword in the module
function getUsedModulesInFunctions(functions?: IFunction[]) {
  if (!functions) return [];

  return functions.flatMap((fun) => {
    const externalFnCalls = fun.externalFnCalls;
    if (!externalFnCalls) return [];
    return externalFnCalls.map(({ namespace, module: name }) => ({
      namespace: namespace || '',
      name,
    }));
  });
}

const isNotDuplicated =
  <T>(isEqual: (a: T, b: T) => boolean) =>
  (a: T, idx: number, list: T[]) =>
    list.findIndex((b) => isEqual(a, b)) === idx;

const isNotDuplicatedModule = isNotDuplicated<{
  name: string;
  namespace?: string;
}>((a, b) => a.name === b.name && a.namespace === b.namespace);

function fileParser(
  contract: string,
  namespace: string = '',
): [IModule[], IPointer] {
  const pointer = getPointer(contract);
  const result = parser(pointer);
  if (result === FAILED) {
    console.error('there is no rule matched with this token', pointer.next());
    throw new Error('parsing error');
  }
  const tree = unwrapData(result);
  const { modules, namespaces, usedModules } = tree;

  const extModules =
    modules?.map(({ location, module: mod }) => ({
      ...mod,
      namespace: getNamespace(location, namespaces) || namespace,
      usedModules: [
        ...(mod.usedModules || []),
        ...getUsedModules(location, usedModules),
        ...getUsedModulesInFunctions(mod.functions),
      ].filter(isNotDuplicatedModule),
    })) || [];

  return [extModules, pointer];
}

// fetch/read the contract file only once
const moduleLoader = (getContract: (fullName: string) => Promise<string>) => {
  const contents: Record<string, Promise<string>> = {};
  return (fullName: string) => {
    if (contents[fullName] === undefined) {
      contents[fullName] = getContract(fullName);
    }
    return contents[fullName];
  };
};

export const getModuleFullName = (mod: { name: string; namespace?: string }) =>
  mod.namespace ? `${mod.namespace}.${mod.name}` : mod.name;

type IModuleWithPointer = IModule & { pointer?: IPointer };

// parse a module and its dependencies
async function parserAllModules(
  fullName: string,
  namespace: string,
  getContent: (name: string) => Promise<string>,
): Promise<Array<IModuleWithPointer>> {
  const content = await getContent(fullName);
  const [modules, pointer] = fileParser(content, namespace);
  const dependencies = await Promise.all(
    modules.map(async (mod) => {
      (mod as any).pointer = pointer;
      const interfaceOrModule = [
        ...(mod.usedModules || []),
        ...(mod.usedInterface || []),
      ];
      if (interfaceOrModule.length === 0) return [];
      const mods = await Promise.all(
        interfaceOrModule.map(async (usedModule) =>
          parserAllModules(
            getModuleFullName(usedModule),
            usedModule.namespace || '',
            getContent,
          ),
        ),
      );
      return mods.flat();
    }),
  );
  return [...modules, ...dependencies.flat()].filter(isNotDuplicatedModule);
}

// check function body again for checking function calls, internal and external
const getFunctionCalls = (
  bodyPointer: number | undefined,
  pointer: IPointer,
  fnList: string[],
  availableExternals: Array<{
    namespace?: string;
    module: string;
    func: string;
  }>,
): {
  internal?: string[];
  external?: Array<{ namespace?: string; module: string; func: string }>;
} => {
  if (bodyPointer !== undefined) {
    pointer.reset(bodyPointer);
    const blockPointer = getBlockPointer(pointer);
    const data = functionCalls(fnList, availableExternals)(blockPointer);
    if (data !== FAILED) {
      return unwrapData(data);
    }
  }

  return {};
};

// adding function calls to each function in the module
function addFunctionCalls(
  mod: IModule,
  allModule: Map<string, IModule>,
  pointer: IPointer,
) {
  if (mod.kind !== 'module') return mod;

  const usedModules = mod.usedModules;
  let externalFunctions: Array<{
    module: string;
    namespace: string | undefined;
    func: string;
  }> = [];

  if (usedModules && usedModules.length) {
    externalFunctions = usedModules.flatMap(({ name, namespace, imports }) => {
      if (imports && imports.length) {
        return imports.map((func) => ({
          module: name,
          namespace,
          func,
        }));
      }
      const usedModule = allModule.get(getModuleFullName({ name, namespace }));
      if (!usedModule || !usedModule.functions) return [];
      return usedModule.functions.map(({ name: func }) => ({
        module: name,
        namespace,
        func,
      }));
    });
  }
  if (!mod.functions) {
    mod.functions = [];
  }
  const internalFunctions = mod.functions.map(({ name }) => name);
  mod.functions = mod.functions.map((fun) => {
    if (!fun.bodyPointer) {
      return fun;
    }
    console.log('body', fun.name, fun.bodyPointer);
    const { internal = [], external = [] } = getFunctionCalls(
      fun.bodyPointer,
      pointer,
      internalFunctions,
      externalFunctions,
    );
    fun.functionCalls = {
      internal,
      external: [...external, ...(fun.externalFnCalls || [])],
    };
    delete fun.bodyPointer;
    return fun;
  });
}

// extract all capabilities for each function in the module
function addFunctionCapabilities(
  mod: IModule,
  allModule: Map<string, IModule>,
) {
  if (mod.kind !== 'module' || !mod.functions) return mod;
  mod.functions.forEach((fun) => {
    fun.allExtractedCaps = getCapabilities(
      allModule,
      getModuleFullName(mod),
      fun.name,
    );
  });
}

export async function pactParser(
  file: string,
  namespace: string = '',
  getContract: (fullName: string) => Promise<string>,
) {
  const modules = await parserAllModules(
    file,
    namespace,
    moduleLoader(getContract),
  );

  const allModules = new Map<string, IModuleWithPointer>();

  modules.forEach((mod) => {
    allModules.set(getModuleFullName(mod), mod);
  });

  modules.forEach((mod) => {
    addFunctionCalls(mod, allModules, mod.pointer!);
    delete mod.pointer;
  });

  modules.forEach((mod) => {
    addFunctionCapabilities(mod, allModules);
  });

  return Object.fromEntries(allModules);
}
