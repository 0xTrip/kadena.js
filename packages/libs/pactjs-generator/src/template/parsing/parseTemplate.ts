/**
 * @alpha
 */
export interface ITemplate {
  parts: TemplateParts;
  holes: TemplateHoles;
}
/**
 * @alpha
 */
export type TemplateHoles = string[];
/**
 * @alpha
 */
export type TemplateParts = string[];

/**
 * @alpha
 */
export function parseTemplate(template: string): ITemplate {
  const [parts, holes] = splitTemplate(template);

  return {
    parts,
    holes,
  };
}

function splitTemplate(template: string): [string[], string[]] {
  return template.split(/(?:\{\{)|(?:\}\})/g).reduce(
    (acc, stringOrHole, i) => {
      // acc[0] is parts
      // acc[1] is holes

      if (i % 2 === 0) {
        // evens are the template parts
        acc[0].push(stringOrHole);
      } else {
        // odds are the holes
        acc[1].push(stringOrHole);
      }
      return acc;
    },
    [[], []] as [string[], string[]],
  );
}

// Pact.templates("@kadena/my-tx-lib").template("something", {hole2: ""})
