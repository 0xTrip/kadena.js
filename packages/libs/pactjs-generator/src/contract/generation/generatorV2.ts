import { IFunction, IModule } from '../parsing/pactParser';
import { getModuleFullName } from '../parsing/utils/utils';

import { EOL } from 'os';

const keywordsMap: Record<string, string> = {
  decimal: 'IPactDecimal',
  integer: 'IPactInt',
  time: 'Date',
  bool: 'boolean',
  guard: '() => string',
};

const mapType = (
  inputType?: string | { kind: string; value: string },
): string => {
  if (inputType === undefined) {
    return 'any';
  }
  if (typeof inputType === 'string') {
    return keywordsMap[inputType] ?? inputType;
  }
  return inputType.kind;
};

const getFuncCapInterfaceName = (func: IFunction): string => {
  if (
    func.allExtractedCaps === undefined ||
    func.allExtractedCaps.length === 0
  ) {
    return '';
  }

  return `ICapability_${func.name.replace(/-/g, '_')}`;
};

const indent = (str: string): string =>
  str
    .split(EOL)
    .map((line) => `  ${line}`)
    .join(EOL);

const getParameters = (
  list?: Array<{
    name: string;
    type:
      | string
      | {
          kind: string;
          value: string;
        };
  }>,
): string[] => {
  if (!list) return [];
  return list.map((arg) => {
    return `${arg.name.replace(/-/g, '')}: ${mapType(arg.type)}`;
  });
};

function genFunCapsInterface(func: IFunction): string {
  if (
    func.allExtractedCaps === undefined ||
    func.allExtractedCaps.length === 0
  ) {
    return '';
  }
  const interfaceName = getFuncCapInterfaceName(func);

  const cap = func.allExtractedCaps.map((cap) => {
    let parameters = [
      `cap: "${cap.fullModuleName}.${cap.name}"`,
      `signer: string`,
    ];
    if (cap.capability.parameters) {
      const args = getParameters(cap.capability.parameters);
      parameters = [...parameters, ...args];
    }
    const comment =
      cap.capability.doc !== undefined
        ? `/**${EOL}* ${cap.capability.doc}${EOL}*/`
        : '';
    const addCap = `addCap(${EOL}${parameters
      .map(indent)
      .join(`, ${EOL}`)}): this`;
    return { comment, addCap };
  });

  const capStr = cap.map((c) => `${c.comment}${EOL}${c.addCap},`).join(EOL);
  return `interface ${interfaceName} {${EOL}${indent(capStr)}${EOL}}`;
}

const getFunctionType = (func: IFunction): string => {
  const capInterfaceName = getFuncCapInterfaceName(func) || 'ICapV2';
  const comment =
    func.doc !== undefined ? `/**${EOL}* ${func.doc}${EOL}*/${EOL}` : '';

  const parameters = getParameters(func.parameters);
  const lnBreak = parameters.length > 1;
  const nl = lnBreak ? EOL : '';
  return indent(
    `${comment}"${func.name}": (${nl}${parameters
      .map((d) => (lnBreak ? indent(d) : d))
      .join(`,${nl}`)}) => Builder<${capInterfaceName}>`,
  );
};

/**
 * @alpha
 */
export function generateDts2(
  moduleFullName: string,
  modules: Record<string, IModule>,
): string {
  const module = modules[moduleFullName];
  if (module === undefined) {
    throw new Error(`Module ${moduleFullName} not found`);
  }

  if (module.functions === undefined) {
    throw new Error(`Module ${moduleFullName} has no functions`);
  }

  const capsInterfaces =
    module.functions?.map(genFunCapsInterface).filter(Boolean).join(EOL) || '';

  const dts = `
import type { ICommandBuilderV2, IPactCommand, ICapV2 } from '@kadena/client';
import type { IPactDecimal, IPactInt } from '@kadena/types';

interface ICapability_Coin_GAS {
  addCap(
    cap: "coin.GAS",
    signer: string
  ): this
}

${capsInterfaces}

type Builder<T> = ICommandBuilderV2 & ICapability_Coin_GAS & T

declare module '@kadena/client' {

  export interface IPactModules {
    ${module.doc ? `/**${EOL}* ${module.doc}${EOL}*/${EOL}` : ''}
    "${getModuleFullName(module)}": {
${indent(indent(module.functions.map(getFunctionType).join(`,${EOL}${EOL}`)))}
    }
  }
}`;
  return dts;
}
