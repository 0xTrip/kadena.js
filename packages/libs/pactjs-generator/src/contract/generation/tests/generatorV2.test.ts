import { pactParser } from '../../parsing/pactParser';
import { generateDts2 } from '../generatorV2';

describe('generateDts2', () => {
  it('return type definition file for a module', async () => {
    const module = `(namespace "user")
      (module test-module governance
        (defcap test-cap (name:string) true)
        (defun test-func:boolean (parameter-one:string parameter-two:boolean )
          (with-capability (test-cap "name"))
        )
      )
    `;

    const modules = await pactParser({
      files: [module],
      getContract: () => Promise.resolve(''),
    });

    const dts = generateDts2('user.test-module', modules);
    expect(dts).toMatchSnapshot();
  });
});
