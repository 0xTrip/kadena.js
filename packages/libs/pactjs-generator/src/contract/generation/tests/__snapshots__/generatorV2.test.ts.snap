// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateDts2 adds module, capability and function docs to the dts file 1`] = `
"
import type { ICommandBuilderV2, IPactCommand, ICapV2 } from '@kadena/client';
import type { IPactDecimal, IPactInt } from '@kadena/types';

interface ICapability_Coin_GAS {
  addCap(
    cap: \\"coin.GAS\\",
    signer: string
  ): this
}

interface ICapability_test_func {
  /**
  * this is defcap doc
  */
  addCap(
    cap: \\"user.test-module.test-cap\\", 
    signer: string, 
    name: string): this,
}

type Builder<T> = ICommandBuilderV2 & ICapability_Coin_GAS & T

declare module '@kadena/client' {

  export interface IPactModules {
    /**
* this is module doc
*/

    \\"user.test-module\\": {
      /**
      * this is defun doc
      */
      \\"test-func\\": (
        parameterone: object,
        parametertwo: boolean) => Builder<ICapability_test_func>
    }
  }
}"
`;

exports[`generateDts2 does not generate capability interface if the function uses no capabilities 1`] = `
"
import type { ICommandBuilderV2, IPactCommand, ICapV2 } from '@kadena/client';
import type { IPactDecimal, IPactInt } from '@kadena/types';

interface ICapability_Coin_GAS {
  addCap(
    cap: \\"coin.GAS\\",
    signer: string
  ): this
}



type Builder<T> = ICommandBuilderV2 & ICapability_Coin_GAS & T

declare module '@kadena/client' {

  export interface IPactModules {
    
    \\"user.test-module\\": {
      \\"test-func\\": (
        parameterone: number,
        parametertwo: boolean) => Builder<ICapV2>
    }
  }
}"
`;

exports[`generateDts2 function parameter is empty if the defun has no param 1`] = `
"
import type { ICommandBuilderV2, IPactCommand, ICapV2 } from '@kadena/client';
import type { IPactDecimal, IPactInt } from '@kadena/types';

interface ICapability_Coin_GAS {
  addCap(
    cap: \\"coin.GAS\\",
    signer: string
  ): this
}



type Builder<T> = ICommandBuilderV2 & ICapability_Coin_GAS & T

declare module '@kadena/client' {

  export interface IPactModules {
    
    \\"user.test-module\\": {
      \\"test\\": () => Builder<ICapV2>
    }
  }
}"
`;

exports[`generateDts2 return type definition file for a module 1`] = `
"
import type { ICommandBuilderV2, IPactCommand, ICapV2 } from '@kadena/client';
import type { IPactDecimal, IPactInt } from '@kadena/types';

interface ICapability_Coin_GAS {
  addCap(
    cap: \\"coin.GAS\\",
    signer: string
  ): this
}

interface ICapability_test_func {
  
  addCap(
    cap: \\"user.test-module.test-cap\\", 
    signer: string, 
    name: string): this,
}

type Builder<T> = ICommandBuilderV2 & ICapability_Coin_GAS & T

declare module '@kadena/client' {

  export interface IPactModules {
    
    \\"user.test-module\\": {
      \\"test-func\\": (
        parameterone: string,
        parametertwo: boolean) => Builder<ICapability_test_func>
    }
  }
}"
`;

exports[`generateDts2 use any type if function parameters dont have a type 1`] = `
"
import type { ICommandBuilderV2, IPactCommand, ICapV2 } from '@kadena/client';
import type { IPactDecimal, IPactInt } from '@kadena/types';

interface ICapability_Coin_GAS {
  addCap(
    cap: \\"coin.GAS\\",
    signer: string
  ): this
}

interface ICapability_test_func {
  
  addCap(
    cap: \\"user.test-module.test-cap\\", 
    signer: string, 
    name: string): this,
}

type Builder<T> = ICommandBuilderV2 & ICapability_Coin_GAS & T

declare module '@kadena/client' {

  export interface IPactModules {
    
    \\"user.test-module\\": {
      \\"test-func\\": (
        parameterone: any,
        parametertwo: boolean) => Builder<ICapability_test_func>
    }
  }
}"
`;

exports[`generateDts2 use object type if function parameters is object{schema} 1`] = `
"
import type { ICommandBuilderV2, IPactCommand, ICapV2 } from '@kadena/client';
import type { IPactDecimal, IPactInt } from '@kadena/types';

interface ICapability_Coin_GAS {
  addCap(
    cap: \\"coin.GAS\\",
    signer: string
  ): this
}

interface ICapability_test_func {
  
  addCap(
    cap: \\"user.test-module.test-cap\\", 
    signer: string, 
    name: string): this,
}

type Builder<T> = ICommandBuilderV2 & ICapability_Coin_GAS & T

declare module '@kadena/client' {

  export interface IPactModules {
    
    \\"user.test-module\\": {
      \\"test-func\\": (
        parameterone: object,
        parametertwo: boolean) => Builder<ICapability_test_func>
    }
  }
}"
`;

exports[`generateDts2 uses the property type if there is no mapped value for that 1`] = `
"
import type { ICommandBuilderV2, IPactCommand, ICapV2 } from '@kadena/client';
import type { IPactDecimal, IPactInt } from '@kadena/types';

interface ICapability_Coin_GAS {
  addCap(
    cap: \\"coin.GAS\\",
    signer: string
  ): this
}



type Builder<T> = ICommandBuilderV2 & ICapability_Coin_GAS & T

declare module '@kadena/client' {

  export interface IPactModules {
    
    \\"user.test-module\\": {
      \\"test\\": (param: newType) => Builder<ICapV2>
    }
  }
}"
`;
